core-backend/sandbox/src/runtime.rs

```diff
@@ -92,80 +99,140 @@ impl<Ext: BackendExternalities> CommonRuntime<Ext> for Runtime<Ext> {
         F: FnOnce(&mut Self) -> Result<T, RunFallibleError>,
         R: From<Result<T, u32>> + Sized,
     {
-        self.run_any(gas, cost, |ctx| {
-            let res = f(ctx);
-            let res = ctx.process_fallible_func_result(res)?;
-
-            // TODO: move above or make normal process memory access.
-            let write_res = ctx.memory_manager.register_write_as::<R>(res_ptr);
-
-            ctx.write_as(write_res, R::from(res)).map_err(Into::into)
-        })
+        self.run_any(
+            gas,
+            cost,
+            |ctx: &mut Self| -> Result<_, UndefinedTerminationReason> {
+                let res = f(ctx);
+                let res = ctx.host_state_mut().process_fallible_func_result(res)?;
+
+                // TODO: move above or make normal process memory access.
+                let write_res = ctx.register_write_as::<R>(res_ptr);
+
+                ctx.write_as(write_res, R::from(res)).map_err(Into::into)
+            },
+        )
     }
 
     fn alloc(&mut self, pages: u32) -> Result<WasmPage, <Ext>::AllocError> {
-        self.ext.alloc(pages, &mut self.memory)
+        let mut state = caller_host_state_take(self.caller);
+        let mut mem = CallerWrap::memory(self.caller, self.memory.clone());
+        let res = state.ext.alloc(pages, &mut mem);
+        self.caller.data_mut().replace(state);
+        res
     }
 }
 
-impl<Ext: BackendExternalities> Runtime<Ext> {
-    // Cleans `memory_manager`, updates ext counters based on globals.
-    fn prepare_run(&mut self, gas: u64) {
-        self.memory_manager = Default::default();
+impl<'a, 'b, Ext: BackendExternalities + 'static> CallerWrap<'a, 'b, Ext> {
+    #[track_caller]
+    pub fn prepare(
+        caller: &'a mut Caller<'b, HostState<Ext, DefaultExecutorMemory>>,
+    ) -> Result<Self, HostError> {
+        let memory = caller
+            .data_mut()
+            .as_ref()
+            .unwrap_or_else(|| unreachable!("Host state must be presented"))
+            .memory
+            .clone();
+        let mut wrapper = Self {
+            caller,
+            manager: Default::default(),
+            memory,
+        };
+
+        let f = || {
+            let gas = wrapper.caller.get_global_val(GLOBAL_NAME_GAS)?;
```

Comment:
why do we need that call? It is redundant if i'm not mistaken

Reply:
prepare fn should accept gas argument, isn't it?